### DISCUSSION TITLE
What's the best way to handle devices running out of space?

### DISCUSSION THREAD

**Discussion by ephread** [onevcat/Kingfisher#1626] [General]
Hey there, thanks for the huge time-saving library!

We've noticed that on devices running out of space, KingFisher would crash with the following error (code: 640 / `NSFileWriteOutOfSpaceError`):

> "You can’t save the file “com.onevcat.Kingfisher.ImageCache.default” because the volume “Data” is out of space."

The incriminated line is [here](https://github.com/onevcat/Kingfisher/blob/master/Sources/Cache/ImageCache.swift#L216)

I can't help but think that if the OS throws `NSFileWriteOutOfSpaceError`, KingFisher should gracefully handle the case by simply disabling any sort of disk caching.

Granted, though, if the device is out of space, most apps will struggle anyway and the phone will be unusable, so it might not be worth it.

What do you think? I'd be willing to work on this, but I wanted to kickstart the discussion first. I didn't see any issues mentioning the topic, so apologies if I missed them.

P. S. I'm not sure whether I should have opened a discussion or an issue, as it's not a bug per se. Let me know, I'm happy to turn this into an issue.

**Comment by onevcat** [comment]
Oh! I just ignored the fact that creating the cache folder also requires disk storage and could be thrown.

I guess it is a so rare case that it gets ignored until now. Yes, we should not do a force try here! I will check how to keep the current public APIs so we can gracefully avoid this situation.

Thanks for reporting this! It would be nice if you kindly create an issue for this topic and I can fix it soon.

**Reply by onevcat** [reply]
Done in #1628 and I still need some time to check the doc/merge and release it.

**Reply by ephread** [reply]
Wow wow, that was a quick update, thanks you so much @onevcat!

> Thanks for reporting this! It would be nice if you kindly create an issue for this topic
I imagine creating the issue is no longer required 😇

Anyway, thanks again for your responsiveness.

**Reply by Binlogo** [reply]
Great! We meet this issue for a while. Thanks for fixing.

**Comment by gg6126811-oss** [comment]
Running out of space is a common issue, and the best approach is to clear unused apps, move large files to cloud storage, and clean temporary data regularly. Keeping backups and checking storage settings helps prevent the problem from happening again.
